go/offline/rpc_cache.go

```diff
@@ -121,40 +126,96 @@ func (c *RPCCache) put(mctx libkb.MetaContext, version Version, rpcName string,
+// Otherwise, `res` should be used by the caller as the response.
+func (c *RPCCache) Serve(mctx libkb.MetaContext, oa keybase1.OfflineAvailability, version Version, rpcName string, encrypted bool, arg interface{}, resPtr interface{},
+	handler func(mctx libkb.MetaContext) (interface{}, error)) (res interface{}, err error) {
 
 	if oa != keybase1.OfflineAvailability_BEST_EFFORT {
-		_, err := handler(mctx)
-		return err
+		return handler(mctx)
 	}
 	mctx = mctx.WithLogTag("OFLN")
 	defer mctx.Trace(fmt.Sprintf("RPCCache#Serve(%d, %s, %v, %+v)", version, rpcName, encrypted, arg), func() error { return err })()
+
+	found, err := c.get(mctx, version, rpcName, encrypted, arg, resPtr)
+	if err != nil {
+		return nil, err
+	}
+
+	// If we know we're not connected, use the cache value right away.
+	// TODO: API calls shouldn't necessarily depend on the
+	// connectivity as measured by gregor.
 	if mctx.G().ConnectivityMonitor.IsConnected(mctx.Ctx()) == libkb.ConnectivityMonitorNo {
-		found, err := c.get(mctx, version, rpcName, encrypted, arg, res)
-		if err != nil {
-			return err
-		}
 		if !found {
-			return libkb.OfflineError{}
+			return nil, libkb.OfflineError{}
 		}
-		return nil
+		return nil, nil // resPtr was filled in by get()
 	}
-	res, err = handler(mctx)
-	if err != nil {
-		return err
+
+	type handlerRes struct {
+		res interface{}
+		err error
 	}
-	tmp := c.put(mctx, version, rpcName, encrypted, arg, res)
-	if tmp != nil {
-		mctx.Warning("Error putting RPC to offline storage: %s", tmp.Error())
+	resCh := make(chan handlerRes, 1)
+
+	// New goroutine needs a new metacontext, in case the original
+	// gets canceled after the timeout below.
+	newMctx := libkb.NewMetaContextBackground(mctx.G())
+	newMctx = newMctx.WithLogTag("OFLN")
+	mctx.Debug("RPCCache#Serve: Launching background best-effort handler with debug tags %s", libkb.LogTagsToString(newMctx.Ctx()))
+
+	// Launch a background goroutine to try to invoke the handler.
+	// Even if we hit a timeout below and return the cached value,
+	// this goroutine will keep going in an attempt to populate the
+	// cache on a slow network.
+	go func() {
+		res, err := handler(newMctx)
+		if err != nil {
+			resCh <- handlerRes{nil, err}
+			return
+		}
+		tmp := c.put(newMctx, version, rpcName, encrypted, arg, res)
+		if tmp != nil {
+			newMctx.Warning("Error putting RPC to offline storage: %s", tmp.Error())
+		}
+		resCh <- handlerRes{res, nil}
+	}()
+
+	var timerCh <-chan time.Time
+	if found {
+		// Use a quick timeout if there's an available cached value.
+		timerCh = mctx.G().Clock().After(bestEffortHandlerTimeout)
+	} else {
+		// Wait indefinitely on the handler if there's nothing in the cache.
+		timerCh = make(chan time.Time)
+	}
+	select {
+	case hr := <-resCh:
+		if hr.err != nil {
+			return nil, hr.err
+		}
+		return hr.res, nil
+	case <-timerCh:
+		mctx.Debug("Timeout waiting for handler; using cached value instead")
```

Comment:
Should we cancel the context passed to `handler` in this case (`newMctx`)?

Reply:
We definitely don't want to cancel the handler's ctx, since the whole point of this goroutine/context shenanigans is to have the handler finish up and populate the cache in the background, even if we serve the request from the cache after the timeout.

Reply:
OK, makes sense.